<a href="https://colab.research.google.com/github/ADLIN-NITHISHA/MARKETING-ANALYSIS/blob/main/CUSTOMER_CHURN_ANALYSIS_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df=pd.read_csv("/content/drive/MyDrive/WA_Fn-UseC_-Telco-Customer-Churn.csv")
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [ ]:
df.shape

(7043, 21)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [ ]:
df = df.drop('customerID', axis=1)
df.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes         No       1           No   
1    Male              0      No         No      34          Yes   
2    Male              0      No         No       2          Yes   
3    Male              0      No         No      45           No   
4  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

DATA PREPROCESSING:

Data preprocessing is a crucial step in model building, involving tasks like handling missing values, scaling features, and encoding categorical variables. This ensures the data is clean, standardized, and ready for effective modeling, ultimately enhancing the performance and accuracy of machine learning algorithms.

In [ ]:
from sklearn.preprocessing import MinMaxScaler
categorical_columns = [col for col in df.columns if df[col].dtype == 'object']
df_encoded = pd.get_dummies(df, columns=categorical_columns)
df_encoded.head()

SeniorCitizen  tenure  MonthlyCharges  gender_Female  gender_Male  \
0              0       1           29.85              1            0   
1              0      34           56.95              0            1   
2              0       2           53.85              0            1   
3              0      45           42.30              0            1   
4              0       2           70.70              1            0   

   Partner_No  Partner_Yes  Dependents_No  Dependents_Yes  PhoneService_No  \
0           0            1              1               0                1   
1           1            0              1               0                0   
2           1            0              1               0                0   
3           1            0              1               0                1   
4           1            0              1               0                0   

   ...  TotalCharges_996.85  TotalCharges_996.95  TotalCharges_997.65  \
0  ...                    0                    0                    0   
1  ...                    0                    0                    0   
2  ...                    0                    0                    0   
3  ...                    0                    0                    0   
4  ...                    0                    0                    0   

   TotalCharges_997.75  TotalCharges_998.1  TotalCharges_999.45  \
0                    0                   0                    0   
1                    0                   0                    0   
2                    0                   0                    0   
3                    0                   0                    0   
4                    0                   0                    0   

   TotalCharges_999.8  TotalCharges_999.9  Churn_No  Churn_Yes  
0                   0                   0         1          0  
1                   0                   0         1          0  
2                   0                   0         0          1  
3                   0                   0         1          0  
4                   0                   0         0          1  

[5 rows x 6577 columns]

In [ ]:
df_encoded['Churn'] = df['Churn']
df_encoded['Churn'] = df_encoded['Churn'].map({'Yes': 1, 'No': 0})
df_encoded = df_encoded.drop(['Churn_No', 'Churn_Yes'], axis=1)
df_encoded = df_encoded.astype(int)

In [ ]:
df_encoded.head()

SeniorCitizen  tenure  MonthlyCharges  gender_Female  gender_Male  \
0              0       1              29              1            0   
1              0      34              56              0            1   
2              0       2              53              0            1   
3              0      45              42              0            1   
4              0       2              70              1            0   

   Partner_No  Partner_Yes  Dependents_No  Dependents_Yes  PhoneService_No  \
0           0            1              1               0                1   
1           1            0              1               0                0   
2           1            0              1               0                0   
3           1            0              1               0                1   
4           1            0              1               0                0   

   ...  TotalCharges_996.45  TotalCharges_996.85  TotalCharges_996.95  \
0  ...                    0                    0                    0   
1  ...                    0                    0                    0   
2  ...                    0                    0                    0   
3  ...                    0                    0                    0   
4  ...                    0                    0                    0   

   TotalCharges_997.65  TotalCharges_997.75  TotalCharges_998.1  \
0                    0                    0                   0   
1                    0                    0                   0   
2                    0                    0                   0   
3                    0                    0                   0   
4                    0                    0                   0   

   TotalCharges_999.45  TotalCharges_999.8  TotalCharges_999.9  Churn  
0                    0                   0                   0      0  
1                    0                   0                   0      0  
2                    0                   0                   0      1  
3                    0                   0                   0      0  
4                    0                   0                   0      1  

[5 rows x 6576 columns]

In [ ]:
x = df_encoded.drop('Churn', axis=1)
y = df_encoded['Churn']

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state= 98)

In [ ]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(max_iter=1000)
logistic_model.fit(x_train, y_train)
y_pred = logistic_model.predict(x_test)

In [ ]:
from sklearn.metrics import accuracy_score, classification_report, precision_recall_fscore_support, precision_recall_curve
accuracy = accuracy_score(y_test, y_pred)
classification_report = classification_report(y_test, y_pred)
print(f"Accuracy: {accuracy}\nClassification Report:\n{classification_report}")

Accuracy: 0.8106956933270232
Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.90      0.88      1595
           1       0.63      0.55      0.59       518

    accuracy                           0.81      2113
   macro avg       0.75      0.72      0.73      2113
weighted avg       0.80      0.81      0.81      2113



INTERPRETATION:

The model exhibits an overall accuracy of 81.07%, demonstrating its effectiveness in correctly predicting customer churn. While precision for the positive class (Churn = 1) is 63%, indicating moderate reliability, the recall of 55% suggests room for improvement in capturing all actual instances of churn.